### ISSUE TITLE
Custom ABK causes sound to drop completely (MW05)

### ISSUE THREAD

**Issue by RAZORXKHARA33** [CrabJournal/ABK_Insert#2]
I was able to successfully port corvette_c5_exh from Carbon over to Most Wanted (both GIN and ABK)... mostly. When I drive a car that uses the sounds, everything works fine--no strange RPM curve or stretching and pitches are all correct. However, if an AI opponent uses a car that uses the new sounds, all sound drops out completely until I'm out of range of that opponent. Then it comes back, until they are back in range, then sound drops out again, rinse and repeat.

I looked at the total size of the custom ABKs and they're almost double that of the ABKs that I started with. This is despite me trimming and downsampling the samples that I pulled from Carbon's ABKs to match the parameters of MW's samples (16-bit mono WAV, 22050 Hz sample rate). I even trimmed the samples to be roughly the same length, though that doesn't seem to have any effect other than reducing the size of the resulting ABK.

It seems to me that the SNRs are simply being appended to the end of the bank, rather than actually replacing the original samples. In fact, I believe that's the intended behavior based on what I'm seeing in the readme and other documentation for this program. That, or some garbage data is being left behind somewhere, and then the sound engine tries to process it and pukes.

I've attached the ABKs I generated and can also attach the WAV files if you want.
[ABKs.zip](https://github.com/CrabJournal/ABK_Insert/files/7989834/ABKs.zip)

CAR_72 is the original, CAR_82 is the custom file. Both EE and SPU are included. I didn't bother modifying SWTN.

UPDATE: I regenerated the ABKs in the same manner as before using the latest version, and now the game just crashes whenever an opponent uses a car that uses the custom sounds. I've attached a different zip containing the ABKs generated with the previous version of ABK_insert, as well as the new ones.
[ABKs_newvsold.zip](https://github.com/CrabJournal/ABK_Insert/files/7990372/ABKs_newvsold.zip)

**Comment by CrabJournal** [comment]
Thank you for report. I just thought for a few days what may went wrong and found issue in sx.exe.

If there's a file with >19 characters name length in same folder as sx.exe, sx.exe just crashing and not generates new xa.raw, abk_insert  using old xa.raw and creates currupted abk file which is causing game crashes.
CAR_82_ENG_MB_SPU_lowrpm.abk for example is 28 character long.
I going to patch sx and make abk_insert delete xa.raw after using it.

Also I have a guess why sound drops when opponent uses a car with new sound, but I need more tests, wait for new release.

**Comment by RAZORXKHARA33** [comment]
So a temporary workaround with the current version is to delete xa.raw after every insert operation?
EDIT: Disregard, I didn't see that there was a new version. Do be aware that the game is still crashing with ABKs generated even with the latest version.

**Comment by RAZORXKHARA33** [comment]
Still crashing with latest version, even with shortened sample names...

**Comment by RAZORXKHARA33** [comment]
So I've gone back to using the old version of ABK_insert, but with the newest version of SX (SoundeXchange) that's shipped with version 3.0. This has given me the chance to port some more sounds over (for the most part). I'm still getting the sound bugging out whenever there's an opponent car using custom ABKs. **However...**

Here's the weird part. If I use a car that uses a sound with custom ABKs, any opponents that also use that car **DO NOT** cause the sound to bug out. In fact, I can hear the samples playing back correctly both for my car and the opponents that use that car. It also does not interfere with other cars that do not use custom ABKs--they play back correctly, too.

**TL;DR,** the sound bugs out for cars that use custom ABKs when the opponent drives such a car and you do not drive that car. However, if you and an opponent drive such a car (i.e., DB9), then the sounds play back correctly (no weird stretching, crackling, etc.) for both of you.

I've attached a zip containing all of the custom ABKs that I have generated thus far, as well as the WAV files. I had to pitch-correct some of them to match the formants in the source ABKs so that the RPM ranges and fade points would be correct. Custom ABKs were made for...

nis_350z_c (car_89)
cor_c5 (car_82, along with lowrpm variant)
ast_db9_a (car_86)
chry_srt8 (car_85)

[sounds.zip](https://github.com/CrabJournal/ABK_Insert/files/8076625/sounds.zip)

The process I'm using to generate these files is as follows:

1) Find a suitable ABK from Most Wanted.
2) Extract wav files from Carbon ABK.
3) Change the formant pitch of samples as needed to match those of the appropriate samples in the original MW ABK (which will be what the new samples get injected into). I also downsample the wavs extracted from the Carbon ABK from 44.1 KHz to 22.05 KHz to match the sample rate of wavs in MW ABKs.
4) Open the MW ABK in a hex editor and change the internal ID to the desired number. To do this, I search for hex value 9CFFFFFF. After the first result is the internal ID of the ABK (in hexadecimal). I also change the filename to reflect this for tracking purposes.
5) Once the wav files are ready and the target ABK is prepared, I run the batch file to generate the new ABK. I do this for both EE and SPU banks. This batch file just automates the process of inserting each wav into the target ABK. To account for the old bug where xa.raw doesn't get deleted, I take care of that in the batch file. This is not needed on the newer versions of ABK_insert, but I'm using the older version since ABKs generated with the newer versions still crash my game.

What I'm noticing is that the resulting ABKs are much larger than what I start with. It looks like the new samples are just being appended to the bank and then the ABK header points to the new samples instead of the old ones. There's still some garbage data being left behind somewhere that's causing the sound engine to freak out and play back a garbled mess. I reckon a better way of doing it would be to clear all of the samples out of the file, then insert, and then change the appropriate offsets in the header (if I'm understanding the structure of ABKs correctly). In this case, you would need to know the number of samples included in the ABK (which we already do, and ABK_insert will tell you that), and then provide the correct number of samples.

**Comment by RAZORXKHARA33** [comment]
@CrabJournal By the way, creating custom ABKs for Carbon+ works flawlessly. I still have to pitch-correct the samples to match those of the original samples (and downsample to 44.1KHz if they are higher than that), but sounds do not cut out at all and the size of the ABK adjusts up or down accordingly. My question then is... why is it still buggy for MW05 and earlier? Should I be indexing my samples from 0 like I have to do with Carbon instead of from 1?

**Comment by CrabJournal** [comment]
I guess problem is small sound memory pool in MW05, the problem is exacerbated by the fact that MW keeps the entire abk file in memory.

**Comment by RAZORXKHARA33** [comment]
> I guess problem is small sound memory pool in MW05, the problem is exacerbated by the fact that MW keeps the entire abk file in memory.

While this is true, it would help if generating new ABKs for MW and earlier behaved the same way as it does for Carbon and up. When I generate a new ABK for Carbon, the file size adjusts according to the total size of all of the samples. That is, if I'm using smaller samples, the size of the resulting ABK will drop (as it should) Likewise, if I'm using larger samples, the size of the ABK will increase (again, as it should). For older formats, regardless of the size of the samples that I'm using, the size of the ABK always increases, often by more than double. It seems to me that ABKs generated for Carbon and up are being properly made (replacing samples rather than adding onto the ABK), while this is not the case for MW and earlier. In other words, garbage data is being left behind with older ABK formats, while the newer one does not leave any behind--this is despite the fact that xa.raw is being deleted properly.

**Comment by RAZORXKHARA33** [comment]
@CrabJournal I was able to trim down the resulting ABK and remove all of the old data, as well as redirect the offsets to the appropriate places (both in ABK header/footer and at the start of BNKL). The sound works perfectly if I drive the car, but if an opponent spawns and it uses a car that uses the custom bank, the game crashes. I am completely lost at this point as to what could be wrong with the file. I also tried doing this with version 3.0, and the game still crashes (both with the file after generation but before trimming, as well as after trimming).
[CAR_82.zip](https://github.com/CrabJournal/ABK_Insert/files/8239236/CAR_82.zip)

**Comment by RAZORXKHARA33** [comment]
I'm beginning to think that there's some hardcoded BS in the game's code that's causing the crashing if the BNKL is modified, or the sound to bug/cut out if the new sound data is appended to the end of the ABK (outside of the BNKL, like in the original version of the ABK_insert tool).

**Comment by RAZORXKHARA33** [comment]
Even with this latest version, which correctly generates the BNKl, the game still crashes when an opponent uses a car with a custom sound (easiest test is the race against Rog from the prologue). I'm really beginning to think there's some hardcoded BS in this game that's causing this.

**Comment by RAZORXKHARA33** [comment]
Almost 4 years later and I think I accidentally found the culprit for the crashing with custom ABKs. However, it wasn't while modding NFSMW I found it. It was when I tried to bring those same ABKs to UG2.

Turns out, if your ABK is too large, then the game will flip out. In UG2, this manifests as the game infinite-loading. In MW05, it crashes. I don't know what the exact limit is, but I've found if I keep my samples roughly the same length as those in the original bank, it actually works. Using the latest version, there are no dropouts when an opponent uses a custom bank (and the custom banks play back correctly, too), and no crashes as long as the ABKs aren't too large. I think the reason the old version would generate banks that kill the game sound was due to the samples being appended to the end of the ABK file, rather than inserted into the BNKL (like your current version does). However, being too large wouldn't cause the game to crash (or infinite load in the case of UG2).

I think this can be marked as solved, and finally I can wrap up and release my sound mods that I had held off on putting out for the last, again, nearly 4 years lol.

TL;DR, crashing is caused by ABKs being too large.
